src/services/id-service.ts

```diff
@@ -0,0 +1,102 @@
+import {NextFunction, Request, response, Response} from 'express';
+import {InvalidEntityError, InvalidRequestError} from '../middlewares/error-handler.js';
+import { closeDBConnection, getTableDefinition, prepareDataSource } from '../middlewares/datasource.js';
+import { Mutex } from 'async-mutex';
+import * as config from '../config.js';
+import {date, RecordType} from "zod";
+import {getRecord} from "../config-validator.js";
+
+const mutex = new Mutex();
+
+// create new id
+async function createId(searchCriteria: {}, entityType: string, next: NextFunction, requestId: number) {
+	console.log('******** CREATE CALLED :' + requestId);
+	const repo = await prepareDataSource(getTableDefinition(entityType), requestId, config.dbSync);
+
+	const savedEntity = await repo.save(searchCriteria);
+	await closeDBConnection(next, requestId).then(() => console.log('DB connection closed'));
+	return savedEntity;
+}
+
+// check if an argo id exists for the submitter and program
+async function findId(searchCriteria: {}, entityType: string, next: NextFunction, requestId: number) {
+	console.log('******** FIND CALLED :' + requestId);
+	const schemaInfo = getTableDefinition(entityType);
+	const repo = await prepareDataSource(schemaInfo, requestId, config.dbSync);
+
+	const keys = Object.keys(searchCriteria) as (keyof typeof searchCriteria)[];
+
+	let query = repo
+		.createQueryBuilder(schemaInfo.tablename)
+		.where(schemaInfo.tablename + '.' + keys[0] + '= :' + keys[0], { [keys[0]]: searchCriteria[keys[0]] });
+
+	if (keys.length > 1) {
+		for (let i = 1; i <= keys.length - 1; i++) {
+			query = query.andWhere(schemaInfo.tablename + '.' + keys[i] + '= :' + keys[i], {
+				[keys[i]]: searchCriteria[keys[i]],
+			});
+		}
+	}
+
+	const entity = await query.addSelect([schemaInfo.tablename + '.entityId']).getOne();
+	await closeDBConnection(next, requestId).then(() => console.log('DB connection closed'));
+	return entity;
+}
+
+export async function getId(request: Request, response: Response, next: NextFunction, requestId: number) {
```

Comment:
The service layer of our application should not have any references to the express route handling. Functions here should perform an action on the server and could be called from any place. Perhaps triggered by a cron job, message queue, http request, gql request, etc. 

The action needs to be independent from the trigger, and this moves all the request handling to the router file.

This function looks to use the following properties from teh request:
- `params.entityType`
- in `getSearchCriteria` it uses all the params?

So with that information we likely want to make a function with the arguments of `entityId` and then all other params as search terms. For return type, let's simplify this for now as returning a string if the ID is found, otherwise undefined indicating not found.

```ts
export async function getId(entityId: string, searchTerms: Record<string, string>): string | undefined { /*...*/}
```

Reply:
Removed all request, response and next references from the service.